File: paimon-common/src/main/java/org/apache/paimon/globalindex/btree/BTreeFileFooter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.globalindex.btree;

import org.apache.paimon.memory.MemorySlice;
import org.apache.paimon.memory.MemorySliceInput;
import org.apache.paimon.memory.MemorySliceOutput;
import org.apache.paimon.sst.BlockHandle;
import org.apache.paimon.sst.BloomFilterHandle;

import javax.annotation.Nullable;

import static org.apache.paimon.utils.Preconditions.checkArgument;

/** The Footer for BTree file. */
public class BTreeFileFooter {

    public static final int MAGIC_NUMBER = 0x50425449;
    public static final int CURRENT_VERSION = 1;
    public static final int ENCODED_LENGTH = 52;

    private final int version;
    @Nullable private final BloomFilterHandle bloomFilterHandle;
    private final BlockHandle indexBlockHandle;
    @Nullable private final BlockHandle nullBitmapHandle;

    public BTreeFileFooter(
            @Nullable BloomFilterHandle bloomFilterHandle,
            BlockHandle indexBlockHandle,
            BlockHandle nullBitmapHandle) {
        this(CURRENT_VERSION, bloomFilterHandle, indexBlockHandle, nullBitmapHandle);
    }

    public BTreeFileFooter(
            int version,
            @Nullable BloomFilterHandle bloomFilterHandle,
            BlockHandle indexBlockHandle,
            BlockHandle nullBitmapHandle) {
        this.version = version;
        this.bloomFilterHandle = bloomFilterHandle;
        this.indexBlockHandle = indexBlockHandle;
        this.nullBitmapHandle = nullBitmapHandle;
    }

    public int getVersion() {
        return version;
    }

    @Nullable
    public BloomFilterHandle getBloomFilterHandle() {
        return bloomFilterHandle;
    }

    public BlockHandle getIndexBlockHandle() {
        return indexBlockHandle;
    }

    @Nullable
    public BlockHandle getNullBitmapHandle() {
        return nullBitmapHandle;
    }

    public static BTreeFileFooter readFooter(MemorySliceInput sliceInput) {
        // read version and verify magic number
        sliceInput.setPosition(ENCODED_LENGTH - 8);

        int version = sliceInput.readInt();
        int magicNumber = sliceInput.readInt();
        checkArgument(
                magicNumber == MAGIC_NUMBER, "File is not a btree index file (bad magic number)");

        sliceInput.setPosition(0);

        // read bloom filter and index handles
        @Nullable
        BloomFilterHandle bloomFilterHandle =
                new BloomFilterHandle(
                        sliceInput.readLong(), sliceInput.readInt(), sliceInput.readLong());
        if (bloomFilterHandle.offset() == 0
                && bloomFilterHandle.size() == 0
                && bloomFilterHandle.expectedEntries() == 0) {
            bloomFilterHandle = null;
        }
        BlockHandle indexBlockHandle = new BlockHandle(sliceInput.readLong(), sliceInput.readInt());

        @Nullable
        BlockHandle nullBitmapHandle = new BlockHandle(sliceInput.readLong(), sliceInput.readInt());
        if (nullBitmapHandle.offset() == 0 && nullBitmapHandle.size() == 0) {
            nullBitmapHandle = null;
        }

        return new BTreeFileFooter(version, bloomFilterHandle, indexBlockHandle, nullBitmapHandle);
    }

    public static MemorySlice writeFooter(BTreeFileFooter footer) {
        MemorySliceOutput output = new MemorySliceOutput(ENCODED_LENGTH);
        writeFooter(footer, output);
        return output.toSlice();
    }

    public static void writeFooter(BTreeFileFooter footer, MemorySliceOutput sliceOutput) {
        // write bloom filter and index handles
        if (footer.bloomFilterHandle == null) {
            sliceOutput.writeLong(0);
            sliceOutput.writeInt(0);
            sliceOutput.writeLong(0);
        } else {
            sliceOutput.writeLong(footer.bloomFilterHandle.offset());
            sliceOutput.writeInt(footer.bloomFilterHandle.size());
            sliceOutput.writeLong(footer.bloomFilterHandle.expectedEntries());
        }

        sliceOutput.writeLong(footer.indexBlockHandle.offset());
        sliceOutput.writeInt(footer.indexBlockHandle.size());

        if (footer.nullBitmapHandle == null) {
            sliceOutput.writeLong(0);
            sliceOutput.writeInt(0);
        } else {
            sliceOutput.writeLong(footer.nullBitmapHandle.offset());
            sliceOutput.writeInt(footer.nullBitmapHandle.size());
        }

        // write version and magic number
        sliceOutput.writeInt(footer.version);
        sliceOutput.writeInt(MAGIC_NUMBER);
    }
}

File: paimon-python/pypaimon/globalindex/btree/btree_file_footer.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
################################################################################

"""B-tree file footer for storing index metadata."""

from dataclasses import dataclass
from typing import Optional
import struct

from pypaimon.globalindex.btree.block_handle import BlockHandle


@dataclass
class BloomFilterHandle:
    """Handle for bloom filter block."""
    
    offset: int
    size: int
    expected_entries: int
    
    def is_null(self) -> bool:
        """Check if this handle represents a null bloom filter."""
        return self.offset == 0 and self.size == 0 and self.expected_entries == 0


class BTreeFileFooter:
    """
    The Footer for BTree file.
    
    This footer contains the handles to the bloom filter, index block, and null bitmap,
    allowing efficient navigation of the B-tree index file.
    """
    
    MAGIC_NUMBER = 198732882
    ENCODED_LENGTH = 48
    
    def __init__(
        self,
        bloom_filter_handle: Optional[BloomFilterHandle],
        index_block_handle: BlockHandle,
        null_bitmap_handle: Optional[BlockHandle]
    ):
        """
        Initialize the BTree file footer.
        
        Args:
            bloom_filter_handle: Handle to the bloom filter block (maybe None)
            index_block_handle: Handle to the index block
            null_bitmap_handle: Handle to the null bitmap block (maybe None)
        """
        self.bloom_filter_handle = bloom_filter_handle
        self.index_block_handle = index_block_handle
        self.null_bitmap_handle = null_bitmap_handle
    
    @classmethod
    def read_footer(cls, data: bytes) -> 'BTreeFileFooter':
        """
        Read footer from byte data.
        
        Args:
            data: Byte data containing the footer
            
        Returns:
            BTreeFileFooter instance
            
        Raises:
            ValueError: If magic number doesn't match
        """
        offset = 0
        
        # Read bloom filter handle
        bf_offset = struct.unpack('<Q', data[offset:offset + 8])[0]
        offset += 8
        bf_size = struct.unpack('<I', data[offset:offset + 4])[0]
        offset += 4
        bf_expected = struct.unpack('<Q', data[offset:offset + 8])[0]
        offset += 8
        
        bloom_filter_handle = None
        if not (bf_offset == 0 and bf_size == 0 and bf_expected == 0):
            bloom_filter_handle = BloomFilterHandle(bf_offset, bf_size, bf_expected)
        
        # Read index block handle
        index_offset = struct.unpack('<Q', data[offset:offset + 8])[0]
        offset += 8
        index_size = struct.unpack('<I', data[offset:offset + 4])[0]
        offset += 4
        
        index_block_handle = BlockHandle(index_offset, index_size)
        
        # Read null bitmap handle
        nb_offset = struct.unpack('<Q', data[offset:offset + 8])[0]
        offset += 8
        nb_size = struct.unpack('<I', data[offset:offset + 4])[0]
        offset += 4
        
        null_bitmap_handle = None
        if not (nb_offset == 0 and nb_size == 0):
            null_bitmap_handle = BlockHandle(nb_offset, nb_size)
        
        # Skip padding
        offset = cls.ENCODED_LENGTH - 4
        
        # Read and verify magic number
        magic_number = struct.unpack('<I', data[offset:offset + 4])[0]
        if magic_number != cls.MAGIC_NUMBER:
            raise ValueError("File is not a table (bad magic number)")
        
        return cls(bloom_filter_handle, index_block_handle, null_bitmap_handle)
